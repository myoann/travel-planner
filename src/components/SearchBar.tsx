"use client";

import { useState, ChangeEvent, KeyboardEvent } from "react";

type TProps = {
  placeholder: string;
  filterBySearch?: (searchQuery: string) => void;
  onSubmit?: (searchQuery: string) => void;
};

export default function SearchBar({
  placeholder,
  filterBySearch,
  onSubmit,
}: TProps) {
  const [searchQuery, setSearchQuery] = useState("");

  /** Handles the search input change */
  const handleSearchChange = (event: ChangeEvent<HTMLInputElement>) => {
    setSearchQuery(event.target.value);

    if (filterBySearch) {
      filterBySearch(event.target.value);
    }
  };

  const handleKeyDown = (event: KeyboardEvent) => {
    if (event.key === "Enter" && onSubmit) {
      onSubmit(searchQuery);
    }
  };

  const handleButtonClick = () => {
    if (onSubmit) {
      onSubmit(searchQuery);
    }
  };

  return (
    <div className="flex items-center w-full h-12 border border-gray-300 rounded shadow-md focus-within:ring focus-within:ring-blue-200">
      <input
        type="text"
        placeholder={placeholder || "Search"}
        className={`flex-grow h-full p-4 focus:outline-none ${
          onSubmit ? "rounded-l" : "rounded"
        }`}
        onChange={handleSearchChange}
        onKeyDown={handleKeyDown}
        value={searchQuery}
      />

      {onSubmit && (
        <button
          className="h-full px-6 bg-blue-500 text-white rounded-r hover:bg-blue-700 focus:outline-none transition duration-300 ease-in-out"
          onClick={handleButtonClick}
        >
          Find it
        </button>
      )}
    </div>
  );
}
