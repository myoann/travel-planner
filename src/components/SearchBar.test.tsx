import { render } from "@testing-library/react";

import SearchBar from "./SearchBar";

const defaultProps = {
  filterBySearch: jest.fn(),
  placeholder: "Search",
};

describe("SearchBar", () => {
  it("matches the snapshot", () => {
    const { container } = render(<SearchBar {...defaultProps} />);

    expect(container).toMatchSnapshot();
  });

  it("renders an input", () => {
    const { container } = render(<SearchBar {...defaultProps} />);

    expect(container.querySelector("input")).toBeInTheDocument();
  });

  it("renders a button if onSubmit is given", () => {
    const { container } = render(
      <SearchBar {...defaultProps} onSubmit={jest.fn()} />,
    );

    expect(container.querySelector("button")).toBeInTheDocument();
  });

  it("does not render a button if onSubmit is not given", () => {
    const { container } = render(<SearchBar {...defaultProps} />);

    expect(container.querySelector("button")).not.toBeInTheDocument();
  });
});
