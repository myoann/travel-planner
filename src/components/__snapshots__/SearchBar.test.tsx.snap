// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar matches the snapshot 1`] = `
<div>
  <div
    class="flex h-12 w-full items-center rounded border border-gray-300 shadow-md focus-within:ring focus-within:ring-blue-200"
  >
    <input
      class="h-full flex-grow p-4 focus:outline-none rounded"
      placeholder="Search"
      type="text"
      value=""
    />
  </div>
</div>
`;
